basalt/src/config/keybinding.rs

```diff
@@ -0,0 +1,182 @@
+
+        Ok(Key {
+            modifiers: parse_modifiers(modifiers).map_err(de::Error::custom)?,
+            code: parse_code(code).map_err(de::Error::custom)?,
+        })
+    }
+}
+
+fn parse_modifiers(modifiers: &str) -> Result<KeyModifiers, KeyParseError> {
+    match modifiers {
+        "" => Ok(KeyModifiers::NONE),
+        "alt" => Ok(KeyModifiers::ALT),
+        "ctrl" | "control" => Ok(KeyModifiers::CONTROL),
+        "hyper" => Ok(KeyModifiers::HYPER),
+        "meta" => Ok(KeyModifiers::META),
+        "shift" => Ok(KeyModifiers::SHIFT),
+        "super" => Ok(KeyModifiers::SUPER),
+        _ => Err(KeyParseError::UnknownModifiers(modifiers.to_string())),
+    }
+}
+fn parse_code(code: &str) -> Result<KeyCode, KeyParseError> {
+    match code.len() {
+        0 => Some(KeyCode::Null),
+        1 => Some(KeyCode::Char(code.chars().next().unwrap())),
+        _ => code
+            .strip_prefix('f')
+            .and_then(|n| n.parse::<u8>().map(KeyCode::F).ok())
+            .or(match code {
+                "backspace" => Some(KeyCode::Backspace),
+                "backtab" => Some(KeyCode::BackTab),
+                "delete" => Some(KeyCode::Delete),
+                "down" => Some(KeyCode::Down),
+                "end" => Some(KeyCode::End),
+                "enter" => Some(KeyCode::Enter),
+                "home" => Some(KeyCode::Home),
+                "insert" => Some(KeyCode::Insert),
+                "left" => Some(KeyCode::Left),
+                "page_down" => Some(KeyCode::PageDown),
+                "page_up" => Some(KeyCode::PageUp),
+                "right" => Some(KeyCode::Right),
+                "tab" => Some(KeyCode::Tab),
+                "up" => Some(KeyCode::Up),
+                _ => None,
+            }),
+    }
+    .ok_or(KeyParseError::UnknownCode(code.to_string()))
+}
+
+#[derive(Debug)]
+enum KeyParseError {
+    InvalidKeybinding(String),
+    UnknownCode(String),
+    UnknownModifiers(String),
+}
+
+impl fmt::Display for KeyParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            KeyParseError::InvalidKeybinding(error) => {
+                write!(f, "Invalid keybinding: {error}")
+            }
+            KeyParseError::UnknownModifiers(modifier) => {
+                write!(f, "Unknown modifiers: {}", modifier)
+            }
+            KeyParseError::UnknownCode(code) => write!(f, "Unknown code: {}", code),
+        }
+    }
+}
+
+impl Error for KeyParseError {}
+
+impl de::Error for KeyParseError {
+    fn custom<T>(msg: T) -> Self
+    where
+        T: fmt::Display,
+    {
+        KeyParseError::InvalidKeybinding(msg.to_string())
+    }
+}
+
+impl From<&KeyEvent> for Key {
+    fn from(
+        KeyEvent {
+            code,
+            modifiers,
+            kind: _,
+            state: _,
+        }: &KeyEvent,
+    ) -> Self {
+        Self {
+            code: *code,
+            modifiers: *modifiers,
+        }
+    }
+}
```

Comment:
I don't see any upside for destructuring here. So would rather simplify it to the following format:
```suggestion
impl From<&KeyEvent> for Key {
    fn from(value: &KeyEvent) -> Self {
        Self {
            code: value.code,
            modifiers: value.modifiers,
        }
    }
}
```

Reply:
I went a little overboard, wanting to account for any addition to crossterm's KeyEvent.
Not quite useful, indeed :sweat_smile: